packages/blocks/plus_minus/src/index.js

```diff
@@ -0,0 +1,1048 @@
+    if (this.itemCount_ == 0) {
+      this.removeInput('EMPTY');
+      this.topInput_ = this.appendValueInput('ADD' + this.itemCount_)
+          .appendField(new FieldPlus(), 'PLUS')
+          .appendField(Blockly.Msg['LISTS_CREATE_WITH_INPUT_WITH']);
+    } else {
+      this.appendValueInput('ADD' + this.itemCount_);
+    }
+    this.itemCount_++;
+  },
+
+  /**
+   * Removes an input from the end of the block. If we are removing the last
+   * input this updates the block to have an 'EMPTY' top input.
+   * @this Blockly.Block
+   * @private
+   */
+  removePart_: function() {
+    this.itemCount_--;
+    this.removeInput('ADD' + this.itemCount_);
+    if (this.itemCount_ == 0) {
+      this.topInput_ = this.appendDummyInput('EMPTY')
+          .appendField(new FieldPlus(), 'PLUS')
+          .appendField(Blockly.Msg['LISTS_CREATE_EMPTY_TITLE']);
+    }
+  },
+
+  /**
+   * Makes it so the minus is visible iff there is an input available to remove.
+   * @private
+   */
+  updateMinus_: function() {
+    const minusField = this.getField('MINUS');
+    if (!minusField && this.itemCount_ > 0) {
+      this.topInput_.insertFieldAt(1, new FieldMinus(), 'MINUS');
+    } else if (minusField && this.itemCount_ < 1) {
+      this.topInput_.removeField('MINUS');
+    }
+  },
+};
+
+/**
+ * Updates the shape of the block to have 3 inputs if no mutation is provided.
+ * @this Blockly.Block
+ * @constructor
+ */
+const listCreateHelper = function() {
+  this.updateShape_(3);
+};
+
+Blockly.Extensions.registerMutator('new_list_create_with_mutator',
+    listCreateMutator, listCreateHelper);
+
+/**
+ * Defines the what are essentially info-getters for the procedures_defnoreturn
+ * block.
+ * @type {{callType_: string, getProcedureDef: (function(): *[])}}
+ */
+const getDefNoReturn = {
+  /**
+   * Returns info about this block to be used by the Blockly.Procedures.
+   * @return {Array} An array of info.
+   * @this Blockly.Block
+   */
+  getProcedureDef: function() {
+    return [this.getFieldValue('NAME'), this.arguments_, false];
+  },
+
+  /**
+   * Used by the context menu to create a caller block.
+   * @type {string}
+   */
+  callType_: 'procedures_callnoreturn',
+};
+
+Blockly.Extensions.registerMixin('get_procedure_def_no_return', getDefNoReturn);
+
+/**
+ * Defines what are essentially info-getters for the procedures_def_return
+ * block.
+ * @type {{callType_: string, getProcedureDef: (function(): *[])}}
+ */
+const getDefReturn = {
+  /**
+   * Returns info about this block to be used by the Blockly.Procedures.
+   * @return {Array} An array of info.
+   * @this Blockly.Block
+   */
+  getProcedureDef: function() {
+    return [this.getFieldValue('NAME'), this.arguments_, true];
+  },
+  /**
+   * Used by the context menu to create a caller block.
+   * @type {string}
+   */
+  callType_: 'procedures_callreturn',
+};
+
+Blockly.Extensions.registerMixin('get_procedure_def_return', getDefReturn);
+
+const procedureContextMenu = {
```

Comment:
I like that you split the procedure code into multiple extensions like this, because it's easy to understand what each extension does.